spotify_player/src/cli/client.rs

```diff
@@ -350,3 +365,425 @@ async fn handle_playback_request(
+    let mut from_ids = Vec::new();
+    for pl_item in from.tracks.items {
+        if let Some(PlayableItem::Track(track)) = pl_item.track {
+            if let Some(id) = track.id {
+                let song = PlaylistData {
+                    id: id.to_owned(),
+                    name: track.name,
+                };
+
+                from_ids.push(song);
+                hasher.write(id.id().as_bytes());
+            }
+        }
+    }
+    let hash = hasher.finish().to_string();
+
+    // Get import dir/file
+    let conf_dir = get_config_folder_path()?;
+    let imports_dir = conf_dir.join("imports");
+    let to_dir = imports_dir.join(import_to.id());
+    let from_file = to_dir.join(import_from.id());
+
+    if !to_dir.exists() {
+        create_dir_all(to_dir)?;
+    }
+
+    if !from_file.exists() {
+        let num_songs = from_ids.len();
+
+        let mut file = OpenOptions::new()
+            .write(true)
+            .create(true)
+            .open(&from_file)?;
+
+        // Write hash and new line to top of file
+        writeln!(file, "{}", hash)?;
+        writeln!(file, "")?;
+
+        for track in from_ids {
+            writeln!(file, "{}:{}", track.id.id(), track.name)?;
+
+            if !to_hash.contains(&track.id) {
+                client
+                    .spotify
+                    .playlist_add_items(import_to.to_owned(), [PlayableId::Track(track.id)], None)
+                    .await?;
+            }
+        }
+        // Possibly create a buffer of tracks that will make for fewer spotify requests
+
+        Ok(format!(
+            "Successfully imported '{}' into '{}'.\n{} songs were added.",
+            from.name, to.name, num_songs
+        ))
+    } else {
+        let file = OpenOptions::new().read(true).open(&from_file)?;
+        let reader = BufReader::new(file);
+        let mut iter = reader.lines();
+
+        let old_hash = iter.next().unwrap()?;
+
+        // Consume empty line
+        iter.next();
+
+        // If old hash doesn't match,
+        // need to do import
+        if !old_hash.eq(&hash) {
+            let mut old_ids = HashSet::new();
+            let mut new_ids = HashSet::new();
+
+            // Write playlists new info to file
+            remove_file(&from_file)?; // Writing wasn't overwriting
+            let mut file = OpenOptions::new()
+                .write(true)
+                .create(true)
+                .open(&from_file)?;
+
+            writeln!(file, "{}", hash)?;
+            writeln!(file, "")?;
+
+            for track in from_ids {
+                let id = track.id.to_owned();
+
+                writeln!(file, "{}:{}", id.id(), track.name)?;
+
+                // Add tracks to hashset
+                new_ids.insert(track);
+
+                if !to_hash.contains(&id) {
+                    client
+                        .spotify
+                        .playlist_add_items(import_to.to_owned(), [PlayableId::Track(id)], None)
```

Comment:
`playlist_add_items` receives a list of ids. So instead of making a single for each id, you can determine the intersection list between `from` playlist and `to` playlist and call `playlist_add_items` with ids from that list.

As you already use `HashSet` to store the track's ids. You can use the `differences` method (https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.difference).

Reply:
I am working on an implementation that uses a buffer of under 100 to send through `playlist_add_items`, just doing the difference may cause a payload larger than the API allows